### PR TITLE
Make Mac, MSW and Unix joysticks behave the same.

### PR DESCRIPTION
**Pull request by mickp**
Addresses trak#18233.

* Moved wxCtz to math.h.
* All implementations use int for m_buttonChange on button events.
* m_buttonChange has one bit set to indicate which button changed.
* Added wxJoystickEvent::GetButtonOrdinal() to return 0-based index of
button that changed.

Perhaps the wxCtz implementation should be moved elsewhere, as it is
currently in the header which may affect compile times.

### CODE REVIEW

File: include/wx/math.h [hunk 1]
```diff
@@ -181,4 +181,24 @@ inline double wxRadToDeg(double rad) { return (rad * 180.0) / M_PI; }
 /* Compute the greatest common divisor of two positive integers */
 WXDLLIMPEXP_BASE unsigned int wxGCD(unsigned int u, unsigned int v);
 
+
+
+/* Count trailing zeros. Use optimised builtin where available. */
```
Comment (vadz) [inline include/wx/math.h]: Duplicate line?

File: include/wx/math.h [hunk 2]
```diff
@@ -181,4 +181,24 @@ inline double wxRadToDeg(double rad) { return (rad * 180.0) / M_PI; }
...
+/* Count trailing zeros. Use optimised builtin where available. */
+inline unsigned int wxCtz(unsigned x)
```
Comment (vadz) [inline include/wx/math.h]: This should be inside `#ifdef __cplusplus` section.

File: interface/wx/event.h [hunk 3]
```diff
@@ -1741,10 +1741,17 @@ class wxJoystickEvent : public wxEvent
...
-        This is a @c wxJOY_BUTTONn identifier, where @c n is one of 1, 2, 3, 4.
+        This was once a @c wxJOY_BUTTONn identifier, where @c n is one of 1, 2, 3, 4,
+        and whose values were 1 << n. To support more than four buttons, the return value
+        is now defined as 1 << n.
     */
     int GetButtonChange() const;
 
+    /**
+        Returns the 0-indexed ordinal of the button changing state.
+    */
```
Comment (vadz) [inline interface/wx/event.h]: Please add `@since 3.1.2` here.

File: include/wx/event.h [hunk 4]
```diff
@@ -2760,6 +2761,7 @@ class WXDLLIMPEXP_CORE wxJoystickEvent : public wxEvent
     int GetZPosition() const { return m_zPosition; }
     int GetButtonState() const { return m_buttonState; }
     int GetButtonChange() const { return m_buttonChange; }
+    int GetButtonOrdinal() const {return wxCtz(m_buttonChange); }
```
Comment (vadz) [inline include/wx/event.h]: Very minor, but please add the missing space after `{`

File: src/osx/core/hidjoystick.cpp [hunk 5]
```diff
@@ -872,7 +872,7 @@ void* wxJoystickThread::Entry()
                 wxevent.SetEventType(wxEVT_JOY_BUTTON_UP);
             }
 
-            wxevent.SetButtonChange(nIndex+1);
+            wxevent.SetButtonChange(1 << nIndex);
```
Comment (vadz) [inline src/osx/core/hidjoystick.cpp]: It might be purely theoretic, but shouldn't we check for overflow here?

**Review by vadz — CHANGES_REQUESTED** [review]
Looks good, but can't be merged before `wxCtz()` is moved into the right place to fix the CI builds.

Could you please do it? TIA!

**Comment by vadz** [comment]
The procedure you describe is correct, not sure what errors are you getting with it.

But I'm not sure we want to add another file just for this... We already define `wxGCD()` in `src/common/utilscmn.cpp`, so we could just define `wxCtz()` there too.

**Comment by mickp** [comment]
Ah - nice one! I was looking for something like a utils source file, but must have skipped over that one.

File: src/osx/core/hidjoystick.cpp (hunk 5 above)
Comment (mickp) [reply src/osx/core/hidjoystick.cpp]: I'll add 
```wxCHECK_RET(j_evt.number < CHAR_BIT * (int) sizeof(int), (void *) -1);```

**Comment by mickp** [comment]
OK - updated with fixes for continuous integration.
* Moved wxCtz implementation from math.h utilscmn.cpp, and renamed to wxCTZ to match style of wxGCD.
* Added bounds checks on bit shifts in joystick implementations.
* Added version note to change in defined behaviour of GetButtonChange.
* Minor style fixes.

File: include/wx/math.h (hunk 2 above)
Comment (vadz) [reply include/wx/math.h]: I'd still put it there because it's not (and shouldn't be) `extern "C"`, so it's not callable from C anyhow.

**Comment by mickp** [comment]
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.

**Comment by mickp** [comment]
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.

**Comment by mickp** [comment]
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.

**Pull request closed without merging**